service/history/api/workflow_id_dedup.go

```diff
@@ -69,7 +73,7 @@ func ResolveDuplicateWorkflowID(
 		case enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING:
 			return nil, ErrUseCurrentExecution
 		case enumspb.WORKFLOW_ID_CONFLICT_POLICY_TERMINATE_EXISTING:
-			return terminateWorkflowAction(newRunID)
+			return resolveJustStartedDuplicatedWorfklows(shardContext, currentWorkflowStartTime, newRunID, workflowID)
```

Comment:
I wonder if we want to enforce the same thing for other policies as well.

User could start & complete a workflowID very fast in a tight loop and we saw before.

Reply:
I was thinking the same. Basically, we don't want the same workflowID to be reused too frequent regardless what policy.

Reply:
While I can see a little backoff for server-started runs like workflow retries and continue-as-new, I would be a bit more hesitant to do this for user-started runs. That a user can do many things in a tight loop we should consider controlling more generally and not specifically for workflow start of the same ID.

Reply:
we are keeping the scope of this task as is.
for other concerns we should have a separate task(s).

and yes:
1) this only applies to the duplicate
2) it only applies if new run is started within small window (<1 sec)

Reply:
👍 
> we are keeping the scope of this task as is.

Can you clarify as-is a bit? So if I start a workflow w/ conflict policy terminate-if-existing and there is one, it will get a 1s penalty, but if there is not one there is no penalty? And if I have workflow retry or CAN, do those impose 1s penalties? Basically anywhere we can document the exact situations the 1s penalty is applied, that would help me understand, thanks!

> and yes:

Can you clarify on point 3 how this affects signal with start? I know signals interrupt start delay. Meaning what happens when I send a signal (or signal with start) during this 1s backoff period, will it be properly set on the new-but-in-1s-backoff workflow or will it fail? Same question for updates and queries and cancellation and such. Just want to understand workflow interactions during this backoff period.

Reply:
Intent is - if you start workflow (and are reusing workflow id):
1. if there is no workflow with the same workflow id, or it is complete - nothing changes
2. if there is workflow with the same workflow id, but it is complete  - nothing changes
3. if there is workflow with the same workflow id running but it started > 1 sec ago (configurable) - nothing changes.
4. (-> main change<- ) But if it was started < 1 sec ago  - it will still run, and we will not start a new one, but  rather return an error, even if policy is "terminate".

it will be the same, signal or not (at least my understanding, but I can be out of my depth here). I will let Yimin weight in.

Reply:
Workflow retry or CAN is not impacted by this PR. 
This PR only impact one case which is when you start a workflow with TerminateIfRunning policy, and there is existing run that is in running state and that was started within 1s. In such case, we reject the new start request. This is to prevent bugs that doing busy loop of start same workflowID over and over again using TerminateIfRunning policy. We have seem this a few times, and each time it was due to bug on user side and they do not intended to do so.